### PR TITLE
Enable all and any for bool tensors

### PR DESCRIPTION
**Pull request by izdeby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21033 Enable all and any for bool tensors**
* #21032 Enabled add, sum, mul and neg for bool tensor
* #21031 Enabled min, max, minall, maxall, cmin, cmax, cmaxValue, cminValue for bool tensors
* #21030 Added possibility to index scalars by bool masks

This PR is a part of a stack which will change result tensor type of comparison ops from uint8 to bool. As this change is rather big and a lot of prep work is needed, im breaking it into a stack.

Changes in this PR:
- Enable all() and any() for bool tensors.

Differential Revision: [D15530497](https://our.internmc.facebook.com/intern/diff/D15530497)

### CODE REVIEW

**Review by colesbury — COMMENTED** [review]
This needs a test for non-empty any/all

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -531,12 +531,12 @@ Tensor all(const Tensor& self) {
...
     self.type().backend() == Backend::CUDA, "all only supports CPU AND CUDA "
     "backend, got: ", toString(self.type().backend()));
-  TORCH_CHECK(self.scalar_type() == at::ScalarType::Byte,
-    "all only supports torch.uint8 dtype");
+  TORCH_CHECK(self.scalar_type() == at::ScalarType::Byte || self.scalar_type() == at::ScalarType::Bool,
+    "all only supports torch.uint8 and torch.bool dtypes");
 
   Tensor result = at::empty({0}, self.options());
   auto iter = make_reduction(
-    "all", result, self, {}, false, at::ScalarType::Byte);
```
Comment (ezyang) [inline aten/src/ATen/native/ReduceOps.cpp]: Good.

**Review by ezyang — APPROVED** [review]
Approved with comments; I agree with Sam, non-empty case needs testing (or better yet, find where we are testing all/any previously and generalize that test code.)

**Review by VitalyFedyunin — APPROVED** [review]

File: aten/src/ATen/native/ReduceOps.cpp (hunk 1 above)
Comment (colesbury) [reply aten/src/ATen/native/ReduceOps.cpp]: This is fine for now, but the explicit return type was intentional. numpy.all and numpy.any always return a `bool` array even for non-bool inputs.

**Pull request closed without merging**